prow/tide/blockers/blockers.go

```diff
@@ -113,11 +118,14 @@ func FindAll(ghc githubClient, log *logrus.Entry, label string, orgRepoTokensByO
 				return
 			}
 			issues = append(issues, result...)
-
+			if useQueryPool {
```

Comment:
I don't think this does what you want it to do. Writing here will block the routine *after* it has touched the GH API, and without a reader you will never hit the `wg.Wait()` call. I suggest that you:

 - use a channel to put all the queries into
 - spawn N goroutines to read from the channel and process until it's done
 
```go
type query struct {
    org string
    query sets.String
}
toQuery := make(chan query, len(queries))
for org, querySet := range queries {
    toQuery <- query{org: org, query: querySet}
}

graphQLRoutines := maxGraphQLGoroutines
if maxGraphQLGoroutines == 0 {
    graphQLRoutines = len(queries)
}
var wg sync.WaitGroup
for i := 0; i < maxGraphQLGoroutines; i++ {
    wg.Add(1)
    go func() {
        defer wg.Done()
        for query := range toQuery {
        // process it
        }
    }
}

wg.Wait()
close(toQuery)
```

Reply:
I am not an expert but it seems to be working at least for me. I engineered this method based on the article: https://go.dev/blog/pipelines
```
func main() {
        max := 3
	blocker := make(chan int, max)
	var wg sync.WaitGroup
	for i := 0; i < 20; i++ {
		blocker <- 1
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			fmt.Println("hello world ", i)
			time.Sleep(time.Second)
			<-blocker
		}(i)
	}
	wg.Wait()
	close(blocker)
}
```
This snippet which is a simplification also works fine. I have ran test version of tide on minikube to see any locks and at least for this small test there were none. Again, maybe I just don't see it (or you meant some specific case), but I think it will work.

Reply:
Sorry, I misread the channel manipulation. This is "clever" code and not IMHO a common Go-ism. I would prefer that we not use these patterns from the blog in 2014 - Go's concurrency has changed a lot over the years and that blog is old enough to entirely predate `context.Context` for managing cancellations.

Please queue data to be processed into a channel and use a bounded set of consumers to ensure concurrency <= N.

Reply:
Ok, just wanted to ask because I did not understand what could possibly go wrong here.